src/linear_algebra/determinant.lean

```diff
@@ -103,4 +104,122 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
 { map_one := det_one,
   map_mul := det_mul }
 
+/-- Transposing a matrix preserves the determinant. -/
+@[simp] lemma det_transpose (M : matrix n n R) : M.transpose.det = M.det :=
+begin
+apply sum_bij (λ σ _, σ⁻¹),
+{ intros σ _, apply mem_univ },
+{ intros σ _,
+  rw [sign_inv],
+  congr' 1,
+  apply prod_bij (λ i _, σ i),
+  { intros i _, apply mem_univ },
+  { intros i _, simp },
+  { intros i j _ _ h, simp at h, assumption },
+  { intros i _, use σ⁻¹ i, finish }
+},
+{ intros σ σ' _ _ h, simp at h, assumption },
+{ intros σ _, use σ⁻¹, finish }
+end
+
+/-- Permuting the columns changes the sign of the determinant. -/
+lemma det_permute (M : matrix n n R) (σ : perm n) : matrix.det (λ i, M (σ i)) = σ.sign * M.det :=
+begin
+unfold det,
+rw mul_sum,
+apply sum_bij (λ τ _, σ * τ),
+{ intros τ _, apply mem_univ },
+{ intros τ _,
+  show
+    ↑(sign τ) * finset.prod univ (λ i, M (σ.to_fun (τ.to_fun i)) i)
+    = ↑(sign σ) * (↑(sign (σ * τ)) * finset.prod univ (λ i, M (σ.to_fun (τ.to_fun i)) i)),
+  rw ←mul_assoc,
+  congr,
+  calc ε τ
+       = ↑(sign σ * sign σ * sign τ) :
+    by {conv_lhs {rw [←one_mul (sign τ), ←int.units_pow_two (sign σ)]}, norm_num}
+  ... = ε σ * ε (σ * τ) :
+    by simp only [mul_assoc, int.cast_mul, sign_mul, coe_coe, units.coe_mul] },
+{ intros τ τ' _ _, exact (mul_left_inj σ).mp },
+{ intros τ _, use σ⁻¹ * τ, use (mem_univ _), exact (mul_inv_cancel_left _ _).symm }
+end
+
+section det_zero
+/-! ### `det_zero` section
+
+  In this section, we give two proofs that a matrix with two identical columns
+  has determinant equal to zero. The first one,
+  `det_zero_of_column_eq_of_char_ne_two` is short but doesn't work in
+  characteristic `2`. The second one, `det_zero_of_column_eq` works in all
+  characteristics, but is more complicated.
+-/
+
+/--
+  `mod_swap i j` contains permutations up to swapping `i` and `j`.
+
+  We use this to partition permutations in the expression for the determinant,
+  such that each partitions sums up to `0`.
+-/
+def mod_swap {n : Type u} [decidable_eq n] (i j : n) : setoid (perm n) :=
+⟨ λ σ τ, σ = τ ∨ σ = swap i j * τ
+, λ σ, or.inl (refl σ)
+, λ σ τ h, or.cases_on h
+(λ h, or.inl h.symm)
+(λ h, or.inr (by rw [h, swap_mul_self_mul]))
+, λ σ τ υ hστ hτυ, by cases hστ; cases hτυ; try {rw [hστ, hτυ, swap_mul_self_mul]}; finish
+⟩
+
+instance (i j : n) : decidable_rel (mod_swap i j).r := λ σ τ, or.decidable
+
+variables {M : matrix n n R} {i j : n} (i_ne_j : i ≠ j) (hij : M i = M j)
+
+include i_ne_j hij
+/-- The determinant is zero if the matrix contains a repeated column.
+
+  The proof shows `M.det = -M.det` and concludes `M.det = 0`,
+  so it doesn't work in characteristic `2`.
+-/
+lemma det_zero_of_column_eq_of_char_ne_two (char_ne_2 : ∀ (a : R), a = -a → a = 0) : M.det = 0 :=
+begin
```

Comment:
I lecture this proof to the undergraduates because it's simpler than the general case -- but do we need two proofs of this in mathlib, one strictly less general than the other?

Reply:
Also: the condition `(char_ne_2 : ∀ (a : R), a = -a → a = 0)` is stronger than `char ≠ 2`. For example, in the ring `Z[X]/(2X)` you'll have `X = - X`, but `X ≠ 0`. Mathlib knows what the characteristic of a ring is: `char_p`. So if you actually want `char ≠ 2`, I guess you could assume `not (char_p R 2)`.

Reply:
> do we need two proofs of this in mathlib, one strictly less general than the other?

True, this was really just an artifact of development that I shouldn't have left in the PR. I've removed the lemma now.